### ISSUE TITLE
Exception: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'elements.canonicalId' in 'where clause'

### ISSUE THREAD

**Issue by brianjhanson** [craftcms/cms#9625]
### Description
Another one, similar to #9624. 

When running migrations from 3.4.X up to 3.7 in the `m200620_230205_field_layout_changes` migration some internal processes (`saveElement` it looks like) assumes that the `canonicalId` column exists, but its migration hasn't run yet.

<details>
<summary>Full Stack trace</summary>

```
*** applying m200620_230205_field_layout_changes
    > add column elements text AFTER `name` to table {{%fieldlayouttabs}} ... done (time: 0.003s)
    > drop column titleLabel from table {{%entrytypes}} ... done (time: 0.003s)
Exception: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'elements.canonicalId' in 'where clause'
The SQL being executed was: SELECT EXISTS(SELECT `elements`.`id`, `elements`.`fieldLayoutId`, `elements`.`uid`, `elements`.`enabled`, `elements`.`archived`, `elements`.`dateCreated`, `elements`.`dateUpdated`, `elements_sites`.`id` AS `siteSettingsId`, `elements_sites`.`slug`, `elements_sites`.`siteId`, `elements_sites`.`uri`, `elements_sites`.`enabled` AS `enabledForSite`, `entries`.`sectionId`, `entries`.`typeId`, `entries`.`authorId`, `entries`.`postDate`, `entries`.`expiryDate`, `content`.`id` AS `contentId`, `content`.`title`, `structureelements`.`root`, `structureelements`.`lft`, `structureelements`.`rgt`, `structureelements`.`level`, `structureelements`.`structureId`, `elements`.`revisionId`, `revisions`.`creatorId` AS `revisionCreatorId`, `revisions`.`num` AS `revisionNum`, `revisions`.`notes` AS `revisionNotes`
FROM (SELECT `elements`.`id` AS `elementsId`, `elements_sites`.`id` AS `elementsSitesId`, `content`.`id` AS `contentId`, `structureelements`.`structureId`
FROM `elements` `elements`
INNER JOIN `entries` `entries` ON `entries`.`id` = `elements`.`id`
INNER JOIN `elements_sites` `elements_sites` ON `elements_sites`.`elementId` = `elements`.`id`
INNER JOIN `content` `content` ON `content`.`elementId` = `elements`.`id`
LEFT JOIN `structureelements` `structureelements` ON (`structureelements`.`elementId` = `elements`.`id`) AND (EXISTS (SELECT *
FROM `structures`
WHERE (`id` = `structureelements`.`structureId`) AND (`dateDeleted` IS NULL)))
INNER JOIN `revisions` `revisions` ON `revisions`.`id` = `elements`.`revisionId`
WHERE (`elements`.`archived`=FALSE) AND (`elements`.`dateDeleted` IS NULL) AND (`elements`.`draftId` IS NULL) AND (`elements`.`canonicalId`=2)
ORDER BY `num` DESC) `subquery`
INNER JOIN `entries` `entries` ON `entries`.`id` = `subquery`.`elementsId`
INNER JOIN `elements` `elements` ON `elements`.`id` = `subquery`.`elementsId`
INNER JOIN `elements_sites` `elements_sites` ON `elements_sites`.`id` = `subquery`.`elementsSitesId`
INNER JOIN `content` `content` ON `content`.`id` = `subquery`.`contentId`
LEFT JOIN `structureelements` `structureelements` ON (`structureelements`.`elementId` = `subquery`.`elementsId`) AND (`structureelements`.`structureId` = `subquery`.`structureId`)
INNER JOIN `revisions` `revisions` ON `revisions`.`id` = `elements`.`revisionId`
ORDER BY `num` DESC) (/home/vagrant/proejct-folder/vendor/yiisoft/yii2/db/Schema.php:678)
#0 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/db/Command.php(1307): yii\db\Schema->convertException()
#1 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/db/Command.php(1168): yii\db\Command->internalExecute()
#2 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/db/Command.php(436): yii\db\Command->queryInternal()
#3 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/db/Query.php(446): yii\db\Command->queryScalar()
#4 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/db/Query.php(213): yii\db\Query->exists()
#5 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/elements/db/ElementQuery.php(1641): craft\db\Query->exists()
#6 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/elements/Entry.php(1544): craft\elements\db\ElementQuery->exists()
#7 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/Elements.php(2457): craft\elements\Entry->beforeSave()
#8 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/Elements.php(773): craft\services\Elements->_saveElementInternal()
#9 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/Sections.php(1536): craft\services\Elements->saveElement()
#10 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/Sections.php(756): craft\services\Sections->_ensureSingleEntry()
#11 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/base/ApplicationTrait.php(1668): craft\services\Sections->handleChangedSection()
#12 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/ProjectConfig.php(1189): craft\console\Application->craft\base\{closure}()
#13 [internal function]: craft\services\ProjectConfig->handleChangeEvent()
#14 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/base/Component.php(628): call_user_func()
#15 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/ProjectConfig.php(752): yii\base\Component->trigger()
#16 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/ProjectConfig.php(1179): craft\services\ProjectConfig->_processConfigChangesInternal()
#17 [internal function]: craft\services\ProjectConfig->handleChangeEvent()
#18 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/base/Component.php(628): call_user_func()
#19 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/ProjectConfig.php(752): yii\base\Component->trigger()
#20 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/services/ProjectConfig.php(521): craft\services\ProjectConfig->_processConfigChangesInternal()
#21 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/migrations/m200620_230205_field_layout_changes.php(87): craft\services\ProjectConfig->set()
#22 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/migrations/m200620_230205_field_layout_changes.php(36): craft\migrations\m200620_230205_field_layout_changes->_updateEntryTypeFieldLayouts()
#23 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/db/Migration.php(52): craft\migrations\m200620_230205_field_layout_changes->safeUp()
#24 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/console/controllers/BaseMigrateController.php(744): craft\db\Migration->up()
#25 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/console/controllers/MigrateController.php(379): yii\console\controllers\BaseMigrateController->migrateUp()
#26 [internal function]: craft\console\controllers\MigrateController->actionAll()
#27 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array()
#28 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/base/Controller.php(181): yii\base\InlineAction->runWithParams()
#29 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/console/Controller.php(184): yii\base\Controller->runAction()
#30 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/base/Module.php(534): yii\console\Controller->runAction()
#31 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/console/Application.php(181): yii\base\Module->runAction()
#32 /home/vagrant/proejct-folder/vendor/craftcms/cms/src/console/Application.php(89): yii\console\Application->runAction()
#33 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/console/Application.php(148): craft\console\Application->runAction()
#34 /home/vagrant/proejct-folder/vendor/yiisoft/yii2/base/Application.php(392): yii\console\Application->handleRequest()
#35 /home/vagrant/proejct-folder/craft(25): yii\base\Application->run()
#36 {main}
*** failed to apply m200620_230205_field_layout_changes (time: 0.076s)
```

</details>

### Steps to reproduce

1. Update from 3.4 to 3.7
2. Attempt to run m200620_230205_field_layout_changes migration.

### Additional info

- Craft version: 3.4.10
- PHP version: 7.4
- Database driver & version: MySQL / 5.7
- Plugins & versions: N/A

**Comment by sjoulbak** [comment]
I have a similar issue here, but then with the upgrade from 2.9.2 to 3.7.4 (`dev-develop as 3.7.4`). The error is thrown at m190913_152146_update_preview_targets. 

<details>
   <summary>Full stack trace</summary>
  
   ```bash
  Exception: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'elements.canonicalId' in 'where clause'
  The SQL being executed was: SELECT EXISTS(SELECT `elements`.`id`, `elements`.`fieldLayoutId`, `elements`.`uid`, `elements`.`enabled`, `elements`.`archived`, `elements`.`dateCreated`, `elements`.`dateUpdated`, `elements_sites`.`id` AS `siteSettingsId`, `elements_sites`.`slug`, `elements_sites`.`siteId`, `elements_sites`.`uri`, `elements_sites`.`enabled` AS `enabledForSite`, `entries`.`sectionId`, `entries`.`typeId`, `entries`.`authorId`, `entries`.`postDate`, `entries`.`expiryDate`, `content`.`id` AS `contentId`, `content`.`title`, `structureelements`.`root`, `structureelements`.`lft`, `structureelements`.`rgt`, `structureelements`.`level`, `structureelements`.`structureId`, `elements`.`revisionId`, `revisions`.`creatorId` AS `revisionCreatorId`, `revisions`.`num` AS `revisionNum`, `revisions`.`notes` AS `revisionNotes`
  FROM (SELECT `elements`.`id` AS `elementsId`, `elements_sites`.`id` AS `elementsSitesId`, `content`.`id` AS `contentId`, `structureelements`.`structureId`
  FROM `craft_elements` `elements`
  INNER JOIN `craft_entries` `entries` ON `entries`.`id` = `elements`.`id`
  INNER JOIN `craft_elements_sites` `elements_sites` ON `elements_sites`.`elementId` = `elements`.`id`
  INNER JOIN `craft_content` `content` ON `content`.`elementId` = `elements`.`id`
  LEFT JOIN `craft_structureelements` `structureelements` ON (`structureelements`.`elementId` = `elements`.`id`) AND (EXISTS (SELECT *
  FROM `craft_structures`
  WHERE `id` = `structureelements`.`structureId`))
  INNER JOIN `craft_revisions` `revisions` ON `revisions`.`id` = `elements`.`revisionId`
  WHERE (`elements`.`archived`=FALSE) AND (`elements`.`draftId` IS NULL) AND (`elements`.`canonicalId`=33)
  ORDER BY `num` DESC) `subquery`
  INNER JOIN `craft_entries` `entries` ON `entries`.`id` = `subquery`.`elementsId`
  INNER JOIN `craft_elements` `elements` ON `elements`.`id` = `subquery`.`elementsId`
  INNER JOIN `craft_elements_sites` `elements_sites` ON `elements_sites`.`id` = `subquery`.`elementsSitesId`
  INNER JOIN `craft_content` `content` ON `content`.`id` = `subquery`.`contentId`
  LEFT JOIN `craft_structureelements` `structureelements` ON (`structureelements`.`elementId` = `subquery`.`elementsId`) AND (`structureelements`.`structureId` = `subquery`.`structureId`)
  INNER JOIN `craft_revisions` `revisions` ON `revisions`.`id` = `elements`.`revisionId`
  ORDER BY `num` DESC) (/app/user/vendor/yiisoft/yii2/db/Schema.php:678)
  #0 /app/user/vendor/yiisoft/yii2/db/Command.php(1307): yii\db\Schema->convertException(Object(PDOException), 'SELECT EXISTS(S...')
  #1 /app/user/vendor/yiisoft/yii2/db/Command.php(1168): yii\db\Command->internalExecute('SELECT EXISTS(S...')
  #2 /app/user/vendor/yiisoft/yii2/db/Command.php(436): yii\db\Command->queryInternal('fetchColumn', 0)
  #3 /app/user/vendor/yiisoft/yii2/db/Query.php(446): yii\db\Command->queryScalar()
  #4 /app/user/vendor/craftcms/cms/src/db/Query.php(213): yii\db\Query->exists(NULL)
  #5 /app/user/vendor/craftcms/cms/src/elements/db/ElementQuery.php(1641): craft\db\Query->exists(NULL)
  #6 /app/user/vendor/craftcms/cms/src/elements/Entry.php(1544): craft\elements\db\ElementQuery->exists()
  #7 /app/user/vendor/craftcms/cms/src/services/Elements.php(2457): craft\elements\Entry->beforeSave(false)
  #8 /app/user/vendor/craftcms/cms/src/services/Elements.php(773): craft\services\Elements->_saveElementInternal(Object(craft\elements\Entry), false, false, NULL)
  #9 /app/user/vendor/craftcms/cms/src/services/Sections.php(1536): craft\services\Elements->saveElement(Object(craft\elements\Entry), false)
  #10 /app/user/vendor/craftcms/cms/src/services/Sections.php(756): craft\services\Sections->_ensureSingleEntry(Object(craft\models\Section), Array)
  #11 /app/user/vendor/craftcms/cms/src/base/ApplicationTrait.php(1668): craft\services\Sections->handleChangedSection(Object(craft\events\ConfigEvent))
  #12 /app/user/vendor/craftcms/cms/src/services/ProjectConfig.php(1189): craft\console\Application->craft\base\{closure}(Object(craft\events\ConfigEvent))
  #13 [internal function]: craft\services\ProjectConfig->handleChangeEvent(Object(craft\events\ConfigEvent))
  #14 /app/user/vendor/yiisoft/yii2/base/Component.php(628): call_user_func(Array, Object(craft\events\ConfigEvent))
  #15 /app/user/vendor/craftcms/cms/src/services/ProjectConfig.php(752): yii\base\Component->trigger('updateItem', Object(craft\events\ConfigEvent))
  #16 /app/user/vendor/craftcms/cms/src/services/ProjectConfig.php(1179): craft\services\ProjectConfig->_processConfigChangesInternal('sections.0de70e...')
  #17 [internal function]: craft\services\ProjectConfig->handleChangeEvent(Object(craft\events\ConfigEvent))
  #18 /app/user/vendor/yiisoft/yii2/base/Component.php(628): call_user_func(Array, Object(craft\events\ConfigEvent))
  #19 /app/user/vendor/craftcms/cms/src/services/ProjectConfig.php(749): yii\base\Component->trigger('addItem', Object(craft\events\ConfigEvent))
  #20 /app/user/vendor/craftcms/cms/src/services/ProjectConfig.php(521): craft\services\ProjectConfig->_processConfigChangesInternal('sections.0de70e...', true, NULL)
  #21 /app/user/vendor/craftcms/cms/src/migrations/m190913_152146_update_preview_targets.php(54): craft\services\ProjectConfig->set('sections.0de70e...', Array)
  #22 /app/user/vendor/craftcms/cms/src/db/Migration.php(52): craft\migrations\m190913_152146_update_preview_targets->safeUp()
  #23 /app/user/vendor/yiisoft/yii2/console/controllers/BaseMigrateController.php(744): craft\db\Migration->up()
  #24 /app/user/vendor/craftcms/cms/src/console/controllers/MigrateController.php(379): yii\console\controllers\BaseMigrateController->migrateUp('m190913_152146_...')
  #25 [internal function]: craft\console\controllers\MigrateController->actionAll()
  #26 /app/user/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
  #27 /app/user/vendor/yiisoft/yii2/base/Controller.php(181): yii\base\InlineAction->runWithParams(Array)
  #28 /app/user/vendor/yiisoft/yii2/console/Controller.php(184): yii\base\Controller->runAction('all', Array)
  #29 /app/user/vendor/yiisoft/yii2/base/Module.php(534): yii\console\Controller->runAction('all', Array)
  #30 /app/user/vendor/yiisoft/yii2/console/Application.php(181): yii\base\Module->runAction('migrate/all', Array)
  #31 /app/user/vendor/craftcms/cms/src/console/Application.php(89): yii\console\Application->runAction('migrate/all', Array)
  #32 /app/user/vendor/yiisoft/yii2/console/Application.php(148): craft\console\Application->runAction('migrate/all', Array)
  #33 /app/user/vendor/yiisoft/yii2/base/Application.php(392): yii\console\Application->handleRequest(Object(craft\console\Request))
  #34 /app/user/craft(27): yii\base\Application->run()
  #35 {main}
  *** failed to apply m190913_152146_update_preview_targets (time: 0.046s)
  ```
</details>

**Comment by brandonkelly** [comment]
I _think_ I’ve just fixed this, however I’m unable to reproduce the error (even with an affected DB backup), so can’t say for sure.

Would be great if one of you could test this quickly and let me know if it works, so we can get an update out with it.

To get the fix early, change your `craftcms/cms` requirement in composer.json to `"dev-develop as 3.7.4"` and run `composer update`.

**Comment by brandonkelly** [comment]
Verified that commit fixed this particular error, and fixed a couple other errors as well (48da2316eb82efe5290bc17ec252c890b8ef10f9 + bee11946fb6ef1c68425836b6d06f3dbbdacf64f). Thanks @brianjhanson!

**Comment by brandonkelly** [comment]
Craft 3.7.5 is out now with those fixes.
